commit cb8dc4e5ce4624a74b941557cdc366425fbd13b9
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Thu Dec 30 02:04:16 2021 -0800

    add support for HZ values for 64, 32, and 16 for DecStation

diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 3ae3d042..29b06019 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -256,6 +256,9 @@ config MACH_DECSTATION
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select SYS_SUPPORTS_64BIT_KERNEL
 	select SYS_SUPPORTS_LITTLE_ENDIAN
+	select SYS_SUPPORTS_16HZ
+	select SYS_SUPPORTS_32HZ
+	select SYS_SUPPORTS_64HZ
 	select SYS_SUPPORTS_128HZ
 	select SYS_SUPPORTS_256HZ
 	select SYS_SUPPORTS_1024HZ
@@ -2566,12 +2569,21 @@ choice
 	help
 	 Allows the configuration of the timer frequency.
 
+	config HZ_16
+		bool "16 HZ" if SYS_SUPPORTS_16HZ || SYS_SUPPORTS_ARBIT_HZ
+
 	config HZ_24
 		bool "24 HZ" if SYS_SUPPORTS_24HZ || SYS_SUPPORTS_ARBIT_HZ
 
+	config HZ_32
+		bool "32 HZ" if SYS_SUPPORTS_32HZ || SYS_SUPPORTS_ARBIT_HZ
+
 	config HZ_48
 		bool "48 HZ" if SYS_SUPPORTS_48HZ || SYS_SUPPORTS_ARBIT_HZ
 
+	config HZ_64
+		bool "64 HZ" if SYS_SUPPORTS_64HZ || SYS_SUPPORTS_ARBIT_HZ
+
 	config HZ_100
 		bool "100 HZ" if SYS_SUPPORTS_100HZ || SYS_SUPPORTS_ARBIT_HZ
 
@@ -2592,12 +2604,22 @@ choice
 
 endchoice
 
+
+config SYS_SUPPORTS_16HZ
+	bool
+
 config SYS_SUPPORTS_24HZ
 	bool
 
+config SYS_SUPPORTS_32HZ
+	bool
+
 config SYS_SUPPORTS_48HZ
 	bool
 
+config SYS_SUPPORTS_64HZ
+	bool
+
 config SYS_SUPPORTS_100HZ
 	bool
 
@@ -2618,8 +2640,11 @@ config SYS_SUPPORTS_1024HZ
 
 config SYS_SUPPORTS_ARBIT_HZ
 	bool
-	default y if !SYS_SUPPORTS_24HZ && \
+	default y if !SYS_SUPPORTS_26HZ && \
+		     !SYS_SUPPORTS_24HZ && \
+		     !SYS_SUPPORTS_32HZ && \
 		     !SYS_SUPPORTS_48HZ && \
+		     !SYS_SUPPORTS_64HZ && \
 		     !SYS_SUPPORTS_100HZ && \
 		     !SYS_SUPPORTS_128HZ && \
 		     !SYS_SUPPORTS_250HZ && \
@@ -2629,8 +2654,11 @@ config SYS_SUPPORTS_ARBIT_HZ
 
 config HZ
 	int
+	default 16 if HZ_16
 	default 24 if HZ_24
+	default 32 if HZ_32
 	default 48 if HZ_48
+	default 64 if HZ_64
 	default 100 if HZ_100
 	default 128 if HZ_128
 	default 250 if HZ_250
diff --git a/arch/mips/kernel/cevt-ds1287.c b/arch/mips/kernel/cevt-ds1287.c
index 77a5ddf5..c5fae94b 100644
--- a/arch/mips/kernel/cevt-ds1287.c
+++ b/arch/mips/kernel/cevt-ds1287.c
@@ -35,6 +35,15 @@ int ds1287_set_base_clock(unsigned int hz)
 	u8 rate;
 
 	switch (hz) {
+	case 16:
+		rate = 0xc;
+		break;
+	case 32:
+		rate = 0xb;
+		break;
+	case 64:
+		rate = 0xa;
+		break;
 	case 128:
 		rate = 0x9;
 		break;
